virtual_rainforest/core/schema.py

```diff
@@ -0,0 +1,151 @@
+        if "default" in subschema:
+            instance.setdefault(property, subschema["default"])
+
+    for error in Draft202012Validator.VALIDATORS["properties"](
+        validator,
+        properties,
+        instance,
+        schema,
+    ):
+        yield error
+
+
+ValidatorWithDefaults = validators.extend(
+    validator=Draft202012Validator, validators={"properties": set_defaults}
+)
+"""A JSONSchema validator that sets defaults where required."""
+
+
+def load_schema(module_name: str, schema_file_path: Path) -> dict:
+    """Function to load the JSON schema for a module.
+
+    This function tries to load a JSON schema file and then - if the JSON loaded
+    correctly - checks that the JSON provides a valid JSON Schema.
+
+    Args:
+        module_name: The name to register the schema under
+        schema_file_path: The file path to the JSON Schema file
+
+    Raises:
+        FileNotFoundError: the schema path does not exist
+        json.JSONDecodeError: the file at the schema path is not valid JSON
+        jsonschema.SchemaError: the file contents are not valid JSON Schema
+        ValueError: the JSON Schema is missing required keys
+    """
+
+    # Try and get the contents of the JSON schema file
+    try:
+        json_schema = json.load(open(schema_file_path))
+    except FileNotFoundError:
+        fnf_error = FileNotFoundError(f"Schema file not found {schema_file_path}.")
+        LOGGER.error(fnf_error)
+        raise fnf_error
+    except json.JSONDecodeError as excep:
+        LOGGER.error(f"JSON error in schema file {schema_file_path}")
+        raise excep
+
+    # Check that this is a valid schema - deliberately log a separate message and let
+    # the schema traceback output rather than replacing it.
+    try:
+        ValidatorWithDefaults.check_schema(json_schema)
+    except exceptions.SchemaError as excep:
+        LOGGER.error(f"Module schema invalid in: {schema_file_path}")
+        raise excep
+
+    # Check that relevant keys are included - deliberately re-raising as ValueError here
+    # as the KeyError has built in quoting and is expecting to provide simple missing
+    # key messages.
+    try:
+        json_schema["properties"][module_name]
+        json_schema["required"]
+    except KeyError as excep:
+        to_raise = ValueError(f"Missing key in module schema {module_name}: {excep}")
+        LOGGER.error(to_raise)
+        raise to_raise
+
+    return json_schema
+
+
+def merge_schemas(schemas: dict[str, dict[str, Any]]) -> dict[str, Any]:
+    """Merge the validation schemas for desired modules.
+
+    The method merges a set of schemas for a set of desired modules into a single
+    integrated schema that can then be used to validate a merged configuration for those
+    modules. The merge also updates the resulting schema to enforce that only properties
+    explicity listed in the schema can be included.
+
+    Args:
+        schema: A dictionary of schemas keyed by module name
```

Comment:
this name is different to the argument (weird as I would have thought the pre-commit hook would catch it)

Reply:
Likewise - that's not good. Seems like something is borked in `pre-commit`? Missing a plugin maybe? @alexdewar @dalonsoa